R/run_sdmtmb.R

```diff
@@ -0,0 +1,224 @@
+	}
+
+	spp <- paste(survey, info[, "scientific_name"], sep = "_")
+
+	# Pull survey data
+	raw_catch_data <- nwfscSurvey::pull_catch(
+		common_name = info$common,
+		survey = survey,
+		dir = file.path(sppdir, survey, "data"))
+	catch_data <- format_data(data = raw_catch_data)
+	catch_data$area <- "coastwide"
+
+	# Rounding is dealing with a WCGBT tow at 31.99861 latitude that was getting dropped
+	if(dim(strata_limits)[1] > 1) {
+		for (a in 2:dim(strata_limits)[1]) {
+			ind <- which(round(catch_data$Lat, 2) < strata_limits[a, "north_border"] & 
+				round(catch_data$Lat, 2) >= strata_limits[a, "south_border"])
+			catch_data$area[ind] <- strata_limits[a, "STRATA"] 
+		}
+	}
+
+	# Summarize tows by year and area
+	cw_tows <- aggregate(Tow ~ Year, catch_data, length, drop = FALSE)
+	area_tows <- aggregate(Tow ~ Year + area, catch_data, length, drop = FALSE)
+	
+
+
+	# Summarize available data
+	surveyspp <- VASTWestCoast::get_spp(input = spp)
+	data_years <- sort(unique(catch_data$Year))
+
+	for (obs in dist) {
+
+		# create subdirectory folder
+		dir.create(file.path(sppdir, survey, "index", obs), recursive = TRUE, showWarnings = FALSE)
+
+		if (obs == "gamma"){
+			family = sdmTMB::delta_gamma()
+		} 
+		if (obs == "lognormal"){
+			family = sdmTMB::delta_lognormal()
+		} 
+		if (!obs %in% c("gamma", "lognormal")){
+			# Add error statement and stop run
+		}
+		
+		# Create mesh
+		mesh_survey <- VASTWestCoast::convert_survey4vast(
+			survey = surveyspp["survey"])
+
+		mesh <- VASTWestCoast::VAST_mesh(
+			data = catch_data,
+		  survey = mesh_survey,
+		  numknots = knots,
+		  savedir = file.path(sppdir, survey, "index", obs))
+		data_mesh <- mesh$mesh$data.inner
+				
+		# Create prediction grid
+		grid <- VASTWestCoast::get_inputgrid(survey = mesh_survey)
+		grid <- grid[grid$Area_km2 > 0, ]
+		grid <- sdmTMB::add_utm_columns(grid, c("Lon", "Lat"), utm_crs = 32610) # UTM 10
+		grid$pass_scaled <- 0
+		grid$vessel_scaled <- data_mesh$vessel_scaled[1]
+		sub_grid <- dplyr::select(grid, X, Y, Area_km2, pass_scaled, vessel_scaled, Lat, Lon)
+		
+		year_grid <- purrr::map_dfr(data_years, function(yr) {
+		  sub_grid$Year <- yr
+		  sub_grid
+		})
+
+		mesh <- sdmTMB::make_mesh(
+    	data = data_mesh,
+    	xy_cols = c("X", "Y"),
+    	n_knots = knots
+	  )
+
+		# plot and save the mesh
+		grDevices::png(filename = file.path(sppdir, survey, "index", obs, "mesh.png"),
+      width = 7, height = 7, units = "in", res = 300, pointsize = 12)
+    plot(mesh)
+    dev.off()
+
+		data_mesh$vessel_scaled <- as.factor(data_mesh$vessel_scaled)
+		
+		# Run model
+		fit <- sdmTMB::sdmTMB(
+  		formula = formula,
+  		time = "Year",
+  		offset = log(data_mesh$Area_swept_km2),
+  		data = data_mesh,
+  		mesh = mesh,
+  		family = family,
+  		spatial = "on",
+  		spatiotemporal = list("iid", "iid"),
+  		anisotropy = TRUE,
+  		silent = TRUE,
+  		control = sdmTMB::sdmTMBcontrol(
+  		  newton_loops = 1L,
```

Comment:
The `newton_loops = 1` and the `ln_H_input` mapping are both defaults now. Not a terrible idea to leave hard coded in case default settings change I guess.

Reply:
Thanks @seananderson I easily see now where newton_loops = 1L is the default but I don't see where the default for a delta model is to set the anisotropy parameters to be shared across the two models, I just see [this](https://github.com/pbs-assess/sdmTMB/blob/3b2424313ae55c26c3b0fd997e2586fa272e98af/R/fit.R#LL1089C18-L1089C50).

Reply:
Not a bad idea to leave both hardcoded in your function just in case the defaults change. I see, for example, that WHAM uses 3 "Newton steps", as people seem to call it, as default. Basically, the `sanity()` check throws gradient errors frequently enough without the additional optimization that I made it the default. It also creates more stability across computer platforms. The `ln_H_input` mapping happens [here](https://github.com/pbs-assess/sdmTMB/blob/3b2424313ae55c26c3b0fd997e2586fa272e98af/R/fit.R#L1312-L1314) if an alternative mapping is not provided.

Reply:
Thanks @seananderson for the link. I think for this two instances I will just go with the defaults because we only needed exact matches when we were in the testing phase, now we can just go with whatever is the best practice. I will make sure though that we throw a `sanity()` check in somewhere as suggested in another comment and here.

Reply:
5a18cc4 and 8ea6c4b resolve this comment.